### DISCUSSION TITLE
Assigning action to a button

### DISCUSSION THREAD

**Discussion by tompad2** [helgoboss/helgobox#882] [Q&A] [answered]
Hi!

Is this possible to do with Realearn (and maybe some lua- scripting)??

I would like to long-press a button on controller and that opens action browser.
When selecting one action and click Run this action is assigned to the same button
but is only run when short-pressing this button. When long-pressing the same
button again the action browser opens and a new action can be assigned to button.

**Comment by helgoboss** [comment, marked as answer]
Yes, to some degree. ReaLearn's learn function also captures executions of actions. This works for some actions better than for others, so you will have to see. I built a small example that uses the target "ReaLearn: Modify mapping - Learn target" to leverage this possibility. The downside is that when picking the action, it will also get executed. If you don't want that, a new feature needs to be built.

Let me know if it works!

Import this into the latest pre-release (2.16.0-pre.1 at the moment), not sure if it would work in older versions.

```lua
{
    kind = "MainCompartment",
    version = "2.16.0-pre.1",
    value = {
        mappings = {
            {
                id = "VvGtPhIWdS3q9xbQL1XXZ",
                name = "Start learn on long press",
                source = {
                    kind = "MidiNoteVelocity",
                    channel = 0,
                    key_number = 64,
                },
                glue = {
                    step_size_interval = {0.01, 0.05},
                    step_factor_interval = {1, 5},
                    fire_mode = {
                        kind = "AfterTimeout",
                        timeout = 1000,
                    },
                },
                target = {
                    kind = "ModifyMapping",
                    mapping = "pW0a8z8xeIjI5b3amuQS_",
                    modification = {
                        kind = "LearnTarget",
                        included_targets = {
                            "RouteVolume",
                            "Tempo",
                            "TrackSoloState",
                            "AutomationModeOverride",
                            "BrowseFxPresets",
                            "ReaperAction",
                            "TrackAutomationMode",
                            "TrackMonitoringMode",
                            "RoutePan",
                            "TrackMuteState",
                            "TrackSelectionState",
                            "TrackArmState",
                            "FxParameterValue",
                            "PlayRate",
                            "TransportAction",
                            "FxOnOffState",
                            "TrackPan",
                            "TrackVolume",
                        },
                    },
                },
            },
            {
                id = "prONxtyySIDrKwF6bwljt",
                name = "Open action browser on long press",
                source = {
                    kind = "MidiNoteVelocity",
                    channel = 0,
                    key_number = 64,
                },
                glue = {
                    step_size_interval = {0.01, 0.05},
                    step_factor_interval = {1, 5},
                    fire_mode = {
                        kind = "AfterTimeout",
                        timeout = 1000,
                    },
                },
                target = {
                    kind = "ReaperAction",
                    command = 40605,
                },
            },
            {
                id = "pW0a8z8xeIjI5b3amuQS_",
                name = "Learned mapping",
                source = {
                    kind = "MidiNoteVelocity",
                    channel = 0,
                    key_number = 64,
                },
                glue = {
                    step_size_interval = {0.01, 0.05},
                },
                target = {
                    kind = "ReaperAction",
                    command = 41042,
                    invocation = "Trigger",
                },
            },
        },
    },
}
```

**Comment by tompad2** [comment]
_The downside is that when picking the action, it will also get executed._ 
Thats not a problem for me :-)

...the part that opens action browser works but not the rest.

This is happening:

I longpress note 64 (E)
The action browser opens
I choose Item: Glue items, ignoring time selection 40362
I press Run/Close and the action is done on two splitted items
I select 2 more items I want to glue.
I shortpress note 64 but something else is happening. 
Looking in Realearns "Learned mapping" the Track:Arm/disarm Track #1 is selected!?

I tried this in MIDI editor also (my first choice to use this) with same result.

**However** then using ReaLearn on a track (not Monitoring FX) it works in Arrange View!! :-)

But not in MIDI Editor :-(


You are right - it will not work with every action. I selected a bounce action
and then running again it run "Set time selection to items".

**Reply by helgoboss** [reply]
MIDI editor? Are you looking for #422? ReaLearn doesn't support MIDI editor actions so far.

Just watch the "Learned mapping". If it suddenly contains action target "Track:Arm/disarm Track #1", that means this action has been executed, it can't come out of nowhere. Maybe you have something else mapped to the button? And remember, some actions are combined of multiple other actions, so ReaLearn can't always pick up the correct action (that's what I meant by "This works for some actions better than for others, so you will have to see."). If you want something non-flaky, ReaLearn would need a new feature. Probably a new mode "Pick REAPER action" for target "ReaLearn: Modify mapping".

**Comment by tompad2** [comment]
> MIDI editor? Are you looking for #422? ReaLearn doesn't support MIDI editor actions so far.

Ahaa - I see! Sorry
> 
> Just watch the "Learned mapping". If it suddenly contains action target "Track:Arm/disarm Track #1", that means this action has been executed, it can't come out of nowhere. Maybe you have something else mapped to the button? 

No I tried with the keyboard keys, all my other mappings is done on XTouch, XTouch mini etc
on Monitoring FX.


>And remember, some actions are combined of multiple other actions, so ReaLearn can't always pick up the correct action (that's what I meant by "This works for some actions better than for others, so you will have to see.").

Like custom actions?

> If you want something non-flaky, ReaLearn would need a new feature. Probably a new mode "Pick REAPER action" for target "ReaLearn: Modify mapping".

That would be nice! :-) I make a request.

Thanks Benjamin for your support.

**Comment by tompad2** [comment]
Hmmm I think I found why Arm/Disarm showed up - I have selected Automatic record arm when track is selected in preferences......
